File: Java/leetcode/src/dataStructure/SinglyLinkedList.java
package dataStructure;

import java.util.StringJoiner;

public class SinglyLinkedList  extends Node{
    private Node head;

    private int size;

    public SinglyLinkedList(){
        head = null;
        size = 0;
    }
    public SinglyLinkedList(Node head,int size){
        this.head = head;
        this.size = size;
    }
    public boolean detectLoop(){
        Node slow = head;
        Node fast = head;
        while(fast!=null && fast.next!=null && fast.next.next!=null){
            fast = fast.next.next;
            slow = slow.next;
            if(slow == fast){
                return true;
            }
        }
        return false;

    }
    public Node middle() {
        if (head == null) {
            return null;
        }
        Node firstCounter = head;
        Node secondCounter = firstCounter.next;
        while (secondCounter != null && secondCounter.next != null) {
            firstCounter = firstCounter.next;
            secondCounter = secondCounter.next.next;
        }
        return firstCounter;
    }

    public void swapNodes(int valueFirst, int valueSecond) {
        if (valueFirst == valueSecond) {
            return;
        }
        Node previousA = null, currentA = head;
        while (currentA != null && currentA.value != valueFirst) {
            previousA = currentA;
            currentA = currentA.next;
        }

        Node previousB = null, currentB = head;
        while (currentB != null && currentB.value != valueSecond) {
            previousB = currentB;
            currentB = currentB.next;
        }
        /** If either of 'a' or 'b' is not present, then return */
        if (currentA == null || currentB == null) {
            return;
        }

        // If 'a' is not head node of list
        if (previousA != null) {
            previousA.next = currentB;
        } else {
            // make 'b' as the new head
            head = currentB;
        }

        // If 'b' is not head node of list
        if (previousB != null) {
            previousB.next = currentA;
        } else {
            // Make 'a' as new head
            head = currentA;
        }
        // Swap next pointer

        Node temp = currentA.next;
        currentA.next = currentB.next;
        currentB.next = temp;
    }

    public Node reverseList(Node node){
        Node prevNode = head;
        while (prevNode.next != null) {
            prevNode = prevNode.next;
        }
        Node prev = null,curr = node ,next;
        while (curr != null) {
            next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }
        prevNode.next = prev;
        return head;
    }
    public void clear(){
        Node cur = head;
        while(cur!=null){
            Node prev = cur;
            cur = cur.next;
            prev = null;
        }
        head = null;
        size = 0;
    }
    public boolean isEmpty(){
        return size == 0;
    }
    public int size(){
        return size;
    }
    public void setHead(Node head){
        this.head = head;
    }
    public int count(){
        int count = 0;
        Node cur = head;
        while(cur!=null){
            cur = cur.next;
            count++;
        }
        return count;
    }
    public boolean search(int value){
        Node cur = head;
        while(cur!=null){
            if(cur.value == value){
                return true;
            }
            cur = cur.next;
        }
        return false;
    }
    @Override
    public String toString(){
        StringJoiner joiner = new StringJoiner("->");
        Node cur = head;
        while (cur!=null){
            joiner.add(cur.value+"->");
            cur = cur.next;
        }
        return joiner.toString();
    }

    public  void deleteDuplicates(){
        Node pred = head;
        Node newHead = head;
        while(pred!=null){
            if (newHead.next !=null && newHead.value == newHead.next.value){
                while (newHead.next !=null && newHead.value == newHead.next.value){
                    newHead = newHead.next;
                }
                pred.next = newHead.next;
                newHead = null;
            }
            pred = pred.next;
            newHead = pred;

        }

    }
    public void print(){
        Node temp = head;
        while (temp!=null && temp.next!=null){
            System.out.println("->");
            temp = temp.next;
        }
        if (temp!=null){
            System.out.println(temp.value);
            System.out.println();
        }
    }
    public void insetHead(int x){
        insertNth(x,0);

    }

    public void insert(int data) {
        insertNth(data, size);
    }

    public void insertNth(int data, int position) {
        checkBounds(position, 0, size);
        Node newNode = new Node(data);
        if (head == null) {
            /* the list is empty */
            head = newNode;
            size++;
            return;
        }
        if (position == 0) {
            /* insert at the head of the list */
            newNode.next = head;
            head = newNode;
            size++;
            return;
        }

        Node cur = head;
        for (int i = 0; i < position - 1; ++i) {
            cur = cur.next;
        }
        newNode.next = cur.next;
        cur.next = newNode;
        size++;
    }

    public void deleteHead() {
        deleteNth(0);
    }

    /**
     * Deletes an element at the tail
     */
    public void delete() {
        deleteNth(size - 1);
    }

    public void deleteNth(int position) {
        checkBounds(position, 0, size - 1);
        if (position == 0) {
            Node destroy = head;
            head = head.next;
            destroy = null;
            /* clear to let GC do its work */
            size--;
            return;
        }
        Node cur = head;
        for (int i = 0; i < position - 1; ++i) {
            cur = cur.next;
        }

        Node destroy = cur.next;
        cur.next = cur.next.next;
        destroy = null; // clear to let GC do its work

        size--;
    }
    public int getNth(int index) {
        checkBounds(index, 0, size - 1);
        Node cur = head;
        for (int i = 0; i < index; ++i) {
            cur = cur.next;
        }
        return cur.value;
    }

    public void checkBounds(int position, int low, int high) {
        if (position > high || position < low) {
            throw new IndexOutOfBoundsException(position + "");
        }
    }




}

File: Python/algoritms/DataStructures/SinglyLinkedList.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

    def __repr__(self):
        return f"Node({self.data})"


class LinkedList:

    def __init__(self):
        self.head: Node | None = None
        self.length = 0

    def __iter__(self):
        node = self.head
        while node:
            yield node
            node = node.next

    def __len__(self):
        return self.length

    def __repr__(self):
        return "->".join(str(node) for node in self)

    def __getitem__(self, index):

        if not 0 <= index < len(self):
            raise IndexError("Index out of range")

        for i, node in enumerate(self):
            if i == index:
                return node
        return None

    def __setitem__(self, index, data):
        if not 0 <= index < len(self):
            raise IndexError("Index out of range")
        current: Node | None = self.head
        for _ in range(index):
            current = current.next
        current.data = data

    def insert_head(self, data):
        self.insert_nth(0, data)

    def insert_nth(self, index, data):
        if not 0 <= index <= len(self):
            raise IndexError("Index out of range")
        new_node = Node(data)
        if self.head is None:
            self.head = new_node
        elif index == 0:
            new_node.next = self.head
            self.head = new_node
        else:
            temp = self.head
            for _ in range(index - 1):
                temp = temp.next
            new_node.next = temp.next
            temp.next = new_node

    def print_list(self):
        print(self)

    def delete_head(self):
        return self.delete_nth(0)

    def delete_tail(self):
        return self.delete_nth(len(self) - 1)

    def delete_nth(self, index):
        if not 0 <= index < len(self):
            raise IndexError("Index out of range")
        deleted_node = self.head
        if index == 0:
            self.head = self.head.next
        else:
            temp = self.head
            for _ in range(index - 1):
                temp = temp.next
            deleted_node = temp.next
            temp.next = temp.next.next

        return deleted_node

    def is_empty(self):
        return self.head is None

    def reverse(self):
        prev = None
        current = self.head
        while current:
            next = current.next
            current.next = prev
            prev = current
            current = next
        self.head = prev


if __name__ == "__main__":
    linkedList = LinkedList()
    linkedList.insert_head(1)
    linkedList.insert_head(2)

    print(len(linkedList))
    print(linkedList)
    print(linkedList.insert_head(5))

    print(linkedList)
